File: concepts/08_async_and_concurrency/05_shared_memory_atomics_and_synchronization/nodejs/test_02_lost_updates_and_predicates.mjs
// 丢失更新、条件谓词与同步边界。
// 共同问题：运行时锁是否自动保护复合操作；条件通知能否替代状态谓词；
// 发布数据需要哪一个明确同步点。
//
// polyglot-family: async_and_concurrency
// polyglot-concept: shared_memory_atomics_and_synchronization
// polyglot-related: languages/nodejs/language/test_030_shared_array_buffer_atomics_and_memory_coordination.mjs

import assert from 'node:assert/strict';
import { once } from 'node:events';
import test from 'node:test';
import { Worker } from 'node:worker_threads';

test('Atomics.wait 和 notify 以共享整数状态发布 payload', async () => {
  const storage = new SharedArrayBuffer(2 * Int32Array.BYTES_PER_ELEMENT);
  const values = new Int32Array(storage);
  const worker = new Worker(`
    const { parentPort, workerData } = require('node:worker_threads');
    const values = new Int32Array(workerData);
    parentPort.postMessage('ready');
    Atomics.wait(values, 0, 0);
    parentPort.postMessage(Atomics.load(values, 1));
  `, { eval: true, workerData: storage });
  const exitPromise = once(worker, 'exit');

  assert.deepEqual(await once(worker, 'message'), ['ready']);
  const resultPromise = once(worker, 'message');
  Atomics.store(values, 1, 42);
  Atomics.store(values, 0, 1);
  Atomics.notify(values, 0, 1);

  assert.deepEqual(await resultPromise, [42]);
  assert.deepEqual(await exitPromise, [0]);

  // 即使 notify 早于 Worker 真正阻塞，wait 也会因状态不再等于 0 返回；状态谓词而非
  // 通知本身防止丢失唤醒。
});

test('Atomics 拒绝非整数 TypedArray', () => {
  const floating = new Float64Array(new SharedArrayBuffer(8));

  assert.throws(() => Atomics.add(floating, 0, 1), TypeError);

  // Atomics 只定义在共享整数/BigInt 视图上；普通对象与浮点视图没有隐式原子化。
});

File: concepts/08_async_and_concurrency/05_shared_memory_atomics_and_synchronization/python/test_02_lost_updates_and_predicates.py
"""丢失更新、条件谓词与同步边界。

共同问题：运行时锁是否自动保护复合操作；条件通知能否替代状态谓词；
发布数据需要哪一个明确同步点。
"""

# polyglot-family: async_and_concurrency
# polyglot-concept: shared_memory_atomics_and_synchronization
# polyglot-related: languages/python/stdlib/088-094_concurrent_execution/
# polyglot-related+: test_088_threading_complete_threads_locks_coordination_introspection_and_low_level_api.py

import threading


def test_coordinated_read_then_write_loses_an_update_without_a_lock():
    value = [0]
    both_read = threading.Barrier(2)

    def increment():
        observed = value[0]
        both_read.wait()
        value[0] = observed + 1

    workers = [threading.Thread(target=increment) for _ in range(2)]
    for worker in workers:
        worker.start()
    for worker in workers:
        worker.join()

    assert value == [1]

    # Barrier 强制两个线程先读同一个旧值再写回；即使 CPython 有 GIL，跨多个操作的
    # 读改写仍会丢失更新。生产代码应以 Lock 包围完整不变量。


def test_condition_rechecks_false_notification_before_proceeding():
    condition = threading.Condition()
    ready = False
    first_check = threading.Event()
    second_check = threading.Event()
    checks = 0
    events = []

    def predicate():
        nonlocal checks
        checks += 1
        if checks == 1:
            first_check.set()
        elif checks == 2:
            second_check.set()
        return ready

    def observe():
        with condition:
            condition.wait_for(predicate)
            events.append("ready")

    worker = threading.Thread(target=observe)
    worker.start()
    first_check.wait()

    with condition:
        condition.notify()
    second_check.wait()
    assert events == []

    with condition:
        ready = True
        condition.notify()
    worker.join()

    assert events == ["ready"]
    assert checks >= 3
